### PR TITLE
modified mnist td-vae to run on mitocheck data 

### PR DESCRIPTION
**Pull request by kmanpearl**
This PR introduces the scripts needed to run train the td-vae model on mitocheck data. `convert_videos.py` (which uses the function in `video_reader.py`) turns preprocessed videos into a single pickle file containing arrays for every movie. `main_train.py`(which uses classes from `model.py`) is used to train the model and save a record of the loss over time as well as the model every 10 epochs. loss can be visualized using `loginfo.py`.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -0,0 +1,4 @@
+__pycache__/*
+loginfo.txt*
+*.pkl
+nohup.out
```
Comment (roshankern) [inline .gitignore]:
```suggestion
nohup.out

```

File: loginfo.py [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+with open("loginfo.txt", "r") as file_handle:
+    for line in file_handle:
+        line = line.strip()
+        field = line.split()
+        if field[-1] != "nan":
+            loss.append(float(field[-1]))
+
+plt.plot(loss[::30])
+plt.ylim(70, 200)
+plt.show()
```
Comment (roshankern) [inline loginfo.py:21]: Consider saving this plot as a png so others viewing the repo can quickly see loss results.

File: loginfo.py [hunk 3]
```diff
@@ -0,0 +1,17 @@
+__author__ = "Keenan Manpearl"
+__date__ = "2023/03/01"
+
```
Comment (roshankern) [inline loginfo.py:3]: Consider adding doc string for this file.

File: main_train.py [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+time_jump_options = [1, 2, 3]  # Jump up to 3 frames away
+
+# Set hyperparameters
+batch_size = 20
+num_epoch = 1000
+learning_rate = 0.0005
+
+data_loader = DataLoader(data, batch_size=batch_size, shuffle=True)
+
+# Build a TD-VAE model
```
Comment (roshankern) [inline main_train.py]: Consider adding documentation to explain your choice for the parameter values.

File: main_train.py [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+
+            print(
+                "epoch: {:>4d}, idx: {:>4d}, loss: {:.2f}".format(
+                    epoch, idx, loss.item()
+                ),
+                file=log_file_handle,
+                flush=True,
+            )
+
+            print(
```
Comment (roshankern) [inline main_train.py]: Consider using [formatted f strings](https://docs.python.org/3/tutorial/inputoutput.html#tut-f-strings) to print this as per the [Way Lab coding conventions](https://docs.google.com/document/d/1htu3DQPZebH2mv-Wyxa45MzUz6Mt8iXZ96t_O0hSSdM/edit#heading=h.e216dnley24u).

File: main_train.py [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+
+        if (epoch + 1) % 10 == 0:
+            torch.save(
+                {
+                    "epoch": epoch,
+                    "model_state_dict": tdvae.state_dict(),
+                    "optimizer_state_dict": optimizer.state_dict(),
+                    "loss": loss,
+                },
+                f"output/model_epoch_{epoch}.pt",
```
Comment (roshankern) [inline main_train.py]: Consider uploading these checkpoints to GitHub if that would be helpful to those viewing the repo.

File: model.py [hunk 7]
```diff
@@ -0,0 +1,625 @@
...
+        self.output_size = output_size
+
+        # Construct fully connected linear layers to map
+        # X to a normal distribution
+        self.weight1 = nn.Linear(input_size, hidden_size)
+        self.weight2 = nn.Linear(input_size, hidden_size)
+        self.mu = nn.Linear(hidden_size, output_size)
+        self.logsigma = nn.Linear(hidden_size, output_size)
+
+    def forward(self, input):
```
Comment (roshankern) [inline model.py:34]: Consider including documentation for what this function does.

File: prep_data.py [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+
+"""
+Create a mitocheck dataset for input into pytorch model
+"""
+import cv2
+import numpy as np
+from torch.utils.data import Dataset
+
+
+class Mitocheck_Dataset(Dataset):
```
Comment (roshankern) [inline prep_data.py:12]: Consider using [numpy docstrings](https://numpydoc.readthedocs.io/en/latest/format.html) to make documentation more clear (example below):

image

File: video_reader.py [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+
+"""
+Reads in an avi video file and returns an array
+"""
+
+import cv2
+import numpy as np
+
+
+def read_video(video_path, n_frames=10):
```
Comment (roshankern) [inline video_reader.py:12]: I think it would be fine to include this short of a function in the file you are using it in, but probably don't change it for this PR.

**Review by roshankern — APPROVED** [review]
A couple general comments:
1) It looks like you don't have any intermediate files included in this PR (loss images, model checkpoint(s). These types of files are usually helpful to those trying to reproduce the repo so please consider adding these!
2) If I understand correctly, it seems like most of these files should be run in a particular order. If this is the case, please consider adding numbers to the files so it is clear other what order they should be run in. Also maybe include this order in the readme or a bash script so it is very clear to others.
3) Please consider doing smaller PRs in the future (although I understand this was a rotation project and you were probably a bit rushed).

Great job overall, feel free to include or ignore my suggestions!

**Comment by roshankern** [comment]
I forgot to mention that its also usually a good idea to create a separate fork in your own repository that you would merge to `WayScience:main` (instead of developing on `kmanpearl:main`)
